core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,247 @@
+  private val scheduler = ThreadUtils.newDaemonSingleThreadScheduledExecutor(
+      "spark-scheduler-blacklist-expire-timer")
+
+  private val recoverPeriod = sparkConf.getTimeAsSeconds(
+    "spark.scheduler.blacklist.recoverPeriod", "60s")
+
+  def start(): Unit = {
+    val scheduleTask = new Runnable() {
+      override def run(): Unit = {
+        Utils.logUncaughtExceptions(expireExecutorsInBlackList())
+      }
+    }
+    scheduler.scheduleAtFixedRate(scheduleTask, 0L, recoverPeriod, TimeUnit.SECONDS)
+  }
+
+  def stop(): Unit = {
+    scheduler.shutdown()
+    scheduler.awaitTermination(10, TimeUnit.SECONDS)
+  }
+
+  // The actual implementation is delegated to strategy
+  /** VisibleForTesting */
+  private[scheduler] def expireExecutorsInBlackList(): Unit = synchronized {
+    val updated = strategy.expireExecutorsInBlackList(executorIdToFailureStatus, clock)
+    if (updated) {
+      invalidAllCache()
+    }
+  }
+
+  def updateFailureExecutors(
+      stageId: Int, info: TaskInfo, reason: TaskEndReason) : Unit = synchronized {
+    val index = info.index
+    val atomTask = BlacklistAtomTask(stageId, index)
+    reason match {
+      // If task succeeding, remove related record from executorIdToFailureStatus
+      case Success =>
+        removeFailureExecutorsForTaskId(info.executorId, stageId, index)
+
+      // If task failing, update latest failure time and failedTaskIds
+      case _ =>
+        val executorId = info.executorId
+        executorIdToFailureStatus.get(executorId) match {
+          case Some(failureStatus) =>
+            failureStatus.updatedTime = clock.getTimeMillis()
+            val failedTimes = failureStatus.numFailuresPerTask.getOrElse(atomTask, 0) + 1
+            failureStatus.numFailuresPerTask.update(atomTask, failedTimes)
+          case None =>
+            val failedTasks = mutable.HashMap(atomTask -> 1)
+            val failureStatus = new FailureStatus(
+              clock.getTimeMillis(), info.host, failedTasks)
+            executorIdToFailureStatus.update(executorId, failureStatus)
+        }
+        invalidAllCache()
+    }
+  }
+
+  /** remove the executorId from executorIdToFailureStatus */
+  def removeFailureExecutors(executorId: String) : Unit = synchronized {
+    executorIdToFailureStatus.remove(executorId)
+    invalidAllCache()
+  }
+
+  /**
+   * remove the failure record related to given taskId from executorIdToFailureStatus. If the
+   * number of records of given executorId becomes 0, remove the completed executorId.
+   */
+  def removeFailureExecutorsForTaskId(
+      executorId: String,
+      stageId: Int,
+      taskIndex: Int) : Unit = synchronized {
+    val atomTask = BlacklistAtomTask(stageId, taskIndex)
+    executorIdToFailureStatus.get(executorId).map(fs => {
+      fs.numFailuresPerTask.remove(atomTask)
+      if(fs.numFailuresPerTask.isEmpty){
+        executorIdToFailureStatus.remove(executorId)
+      }
+      invalidAllCache()
+    })
```

Comment:
nit: you can just use `{}` for `map`, also spaces around the condition for `if`, so

```scala
executorIdToFailureStatus.get(executorId).map { fs => 
  fs.numFailuresPerTask.remove(atomTask)
  if (fs.numFailuresPerTask.isEmpty) {
    executorIdToFailureStatus.remove(executorId)
  }
  invalidAllCache()
  }
```